app/src/main/java/com/mailerdaemon/app/campusMap/MapsActivity.kt

```diff
@@ -0,0 +1,235 @@
+                    val location = locationList[locationList.size - 1]
+                    mLastLocation = location
+                    mCurrLocationMarker.remove()
+
+                    //Place current location marker
+                    val latLng = LatLng(location.latitude, location.longitude)
+                    val markerOptions = MarkerOptions()
+
+                    markerOptions.let {
+                        it.position(latLng)
+                        it.title(resources.getString(R.string.marker_title))
+                        it.icon(BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_MAGENTA))
+                    }
+                    mCurrLocationMarker = mMap.addMarker(markerOptions)
+                }
+            }
+        }
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            if (ContextCompat.checkSelfPermission(this,
+                            Manifest.permission.ACCESS_FINE_LOCATION)
+                    == PackageManager.PERMISSION_GRANTED) {
+                //Location Permission already granted
+                mFusedLocationClient.requestLocationUpdates(mLocationRequest, mLocationCallback, Looper.myLooper())
+                mMap.isMyLocationEnabled = true
+            } else {
+                //Request Location Permission
+                checkLocationPermission()
+            }
+        } else {
+            mFusedLocationClient.requestLocationUpdates(mLocationRequest, mLocationCallback, Looper.myLooper())
+            mMap.isMyLocationEnabled = true
+        }
+    }
+
+    private fun checkLocationPermission() {
+        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
+                != PackageManager.PERMISSION_GRANTED) {
+
+            // Should we show an explanation?
+            if (ActivityCompat.shouldShowRequestPermissionRationale(this,
+                            Manifest.permission.ACCESS_FINE_LOCATION)) {
+
+                // Show an explanation to the user *asynchronously* -- don't block
+                // this thread waiting for the user's response! After the user
+                // sees the explanation, try again to request the permission.
+
+                val builder = AlertDialog.Builder(this)
+                with(builder)
+                {
+                    setTitle(resources.getString(R.string.dialog_title))
+                    setMessage(resources.getString(R.string.dialog_message))
+                    setPositiveButton("OK") { dialogInterface: DialogInterface?, i: Int ->
+                        //Prompt the user once explanation has been shown
+                        ActivityCompat.requestPermissions(this@MapsActivity, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
+                                MY_PERMISSIONS_REQUEST_LOCATION)
+                    }
+                }
+                val alertDialog:AlertDialog = builder.create()
+                alertDialog.show()
```

Comment:
you can actually combine 178 to 190 in a more idiomatic way. use `run{ }.show()`. Or let it be `with` but don't need the variable `builder`.